distributed/tests/test_client.py

```diff
@@ -1116,6 +1117,27 @@ def g():
     result = yield y._result()
     assert result == 456
 
+@gen_cluster(client=True)
+def test_upload_file_zip(c, s, a, b):
+    def g():
+        import myfile
+        return myfile.f()
+
+    with tmp_text('myfile.py', 'def f():\n    return 123') as fn_my_file, \
+         tmp_text('init.py', '') as fn_init:
+        try:
+            with zipfile.ZipFile('myfile.zip', 'w') as z:
+                for fn in (fn_my_file, fn_init):
+                    z.write(fn, arcname=os.path.basename(fn))
+            yield c._upload_file('myfile.zip')
+        finally:
+            if os.path.exists('myfile.zip'):
+                os.remove('myfile.zip')
+
+    sleep(1)  # TODO:  why is this necessary?
```

Comment:
I added the invalid_caches and commonalized some code:
 
```
try:
                name, ext = os.path.splitext(filename)
                names_to_import = []
                if ext in ('.py', '.pyc'):
                    names_to_import.append(name)
                if ext in ('.egg', '.zip'):
                    if out_filename not in sys.path:
                        sys.path.insert(0, out_filename)
                    if ext == '.egg':
                        import pkg_resources
                        pkgs = pkg_resources.find_distributions(out_filename)
                        for pkg in pkgs:
                            names_to_import.append(pkg.project_name)
                    elif ext == '.zip':
                        names_to_import.append(name)
                
                if not names_to_import:
                    logger.warning("Found nothing to import from %s", filename)
                else:
                    invalidate_caches()
                    for name in names_to_import:
                        logger.info("Reload module %s from %s file", name, ext)
                        reload(import_module(name))
```
I removed the sleep in my test case and it works fine (But it also works without the call to invalidate_caches).
However the test still fails in the 'test_upload_file' case which only uploads .py or .pyc files